pkg/stack/unwind/executable.go

```diff
@@ -16,13 +16,73 @@ package unwind
 
 import (
 	"debug/elf"
+	"errors"
 	"fmt"
+	"os"
 	"strings"
+	"syscall"
 
+	burrow "github.com/goburrow/cache"
 	"github.com/hashicorp/go-version"
 	"github.com/xyproto/ainur"
 )
 
+type FramePointerCache struct {
+	cache burrow.Cache
+}
+
+// the inode value can be recycled (this is implementation specific)
+// and it's only unique across filesystems. By adding the creation time
+// to the cache key we significantly reduce the chances of producing wrong
+// results.
+type framePointerCacheKey struct {
+	inode        uint64
+	creationTime syscall.Timespec
+}
+
+func (fpc *FramePointerCache) cacheKey(executable string) (framePointerCacheKey, error) {
+	fileinfo, err := os.Stat(executable)
+	if err != nil {
+		return framePointerCacheKey{}, err
+	}
+
+	stat, ok := fileinfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return framePointerCacheKey{}, errors.New("fileinfo didn't have stat_t")
+	}
+
+	return framePointerCacheKey{
+		inode:        stat.Ino,
+		creationTime: stat.Ctim,
+	}, nil
+}
+
+func (fpc *FramePointerCache) HasFramePointers(executable string) (bool, error) {
+	cacheKey, err := fpc.cacheKey(executable)
+	if err != nil {
+		return false, err
+	}
+
+	val, found := fpc.cache.GetIfPresent(cacheKey)
+	if found {
+		return val.(bool), nil
+	}
+
+	val, err = HasFramePointers(executable)
+	if err != nil {
+		return false, err
+	}
+
+	fpc.cache.Put(cacheKey, val)
+	return false, nil
+}
+
+func NewHasFramePointersCache() FramePointerCache {
+	return FramePointerCache{
+		cache: burrow.New(burrow.WithMaximumSize(10_000)), // 3 * 64bit * 10k (excluding metadata)
```

Comment:
Note to self: check the expiry policies and select one that works for us

Reply:
Sorry, this was just a quick draft, it was the size of the value of the hashmap alone, there is 1 `uint64` and 2 `int64` from `syscall.Timespec`. I'm going to update this with more accurate numbers